src/hotspot/share/gc/parallel/parallelScavengeHeap.cpp

```diff
@@ -539,6 +539,79 @@ void ParallelScavengeHeap::object_iterate(ObjectClosure* cl) {
   old_gen()->object_iterate(cl);
 }
 
+void ParallelScavengeHeap::object_iterate_parallel(ObjectClosure* cl,
+                                                   uint worker_id,
+                                                   HeapBlockClaimer* claimer) {
+  uint block_index;
+  // Iterate until all blocks are claimed
+  while (claimer->claim_and_get_block(&block_index)) {
+    if (block_index == HeapBlockClaimer::eden_index) {
+      young_gen()->eden_space()->object_iterate(cl);
+    } else if (block_index == HeapBlockClaimer::survivor_index) {
+      young_gen()->from_space()->object_iterate(cl);
+      young_gen()->to_space()->object_iterate(cl);
+    } else {
+      uint index = block_index - HeapBlockClaimer::num_inseparable_spaces;
+      old_gen()->block_iterate(cl, index);
+    }
+  }
+}
+
+HeapBlockClaimer::HeapBlockClaimer(uint n_workers) :
+    _n_workers(n_workers), _n_blocks(0), _claims(NULL) {
+  assert(n_workers > 0, "Need at least one worker.");
+  size_t old_gen_used = ParallelScavengeHeap::heap()->old_gen()->used_in_bytes();
+  size_t block_size = ParallelScavengeHeap::heap()->old_gen()->iterate_block_size();
+  uint n_blocks_in_old = old_gen_used / block_size + 1;
+  _n_blocks = n_blocks_in_old + num_inseparable_spaces;
+  _unclaimed_index = 0;
+  uint* new_claims = NEW_C_HEAP_ARRAY(uint, _n_blocks, mtGC);
+  memset(new_claims, Unclaimed, sizeof(*_claims) * _n_blocks);
+  _claims = new_claims;
+}
+
+HeapBlockClaimer::~HeapBlockClaimer() {
+   FREE_C_HEAP_ARRAY(uint, _claims);
+}
+
+bool HeapBlockClaimer::claim_and_get_block(uint* block_index) {
+  assert(block_index != NULL, "Invalid index pointer");
+  uint next_index = Atomic::load(&_unclaimed_index);
+  while (true) {
+    if (next_index >= _n_blocks) {
+      return false;
+    }
+    uint old_val = Atomic::cmpxchg(&_claims[next_index], Unclaimed, Claimed);
+    if (old_val == Unclaimed) {
+      *block_index = next_index;
+      Atomic::inc(&_unclaimed_index);
+      return true;
+    }
+    next_index = Atomic::load(&_unclaimed_index);
+  }
```

Comment:
I think this can be simplified a bit. If you use Atomic::fetch_and_add() for the claiming, you only need `_unclaimed_index` and can get rid of `_claims`. I think you might want to rename it to `_claimed_index` or something. It could look something like this:
```suggestion
  assert(block_index != NULL, "Invalid index pointer");
  *block_index = Atomic::fetch_and_add(&_claimed_index, 1);
  if (*block_index  >= _n_blocks) {
    return false;
  }
  return true;  
```